### ISSUE TITLE
Chained Schedulers

### ISSUE THREAD

**Issue by ravi-mosaicml** [mosaicml/composer#645]
Currently, if one wants to have a warmup or cooldown period, one has to write a custom scheduler for that. Instead, we could have an abstraction for supporting multiple scheduling periods. This could be done via 1) a `ChainedScheduler`-like approach, 2) applying multiple schedulers simultaneously, or 3) making scheduling periods a first-class feature of the trainer.

I am leaning against option 3 because  it would overly complicate the trainer for the generic use case where one only has one training period. While option 2 would be syntactically simpler, it would introduce challenges with SSR. As such, I am leaning towards option 1.

For option 1:
* Each ComposerScheduler should take an `apply_ssr` parameter (which can be set to False if being used in warmup) and an `end_time` or `period_length` parameter (or both, but only allow one to be set). A scheduler should not assume a `start_time` -- that would be determined implicitly whenever  the scheduler is first `__call__`ed.
* We would also need to modify the `__call__` API of the ComposerScheduler such that a scheduler should return `None` when it is finished (i.e. its `end_time` or `period_length` has been elapsed). None would signal to a `ChainedScheduler` that the scheduler is done and is no longer managing the learning rate. If the trainer received `None`, it would interpret that as to not modify the learning rate (which would be equivalent to returning the last returned value). For schedulers (e.g. cooldown) that should run until training end, the `end_time` parameter` could be `1dur`, in which case it would never return None.
* We can have a `ChainedScheduler(schedulers: List[Scheduler] | OrderedDict[str, Scheduler])`-like API . (A dict could be used to give each period a name -- e.g. "warmup"). Whenever the `ChainedScheduler` is `__call__`ed, it would `__call__` the 0th scheduler until it returns None; then it would move on to the 1st scheduler until that returns None, etc. After all schedulers would be exhausted, it would return `None`, which would signal to the trainer to leave the learning rate the same.
* The `ChainedScheduler` would make available the currently active scheduler as an object. Assuming only one ChainedScheduler is being used, algorithms could inspect `state.schedulers[0].active_scheduler`, `state.schedulers[0].active_scheduler_idx`, or `state.schedulers[0].active_scheduler_name` to determine which learning rate period we are in.

Thoughts?

CC @dblalock @jbloxham  @mosaicml/composer-team-research

_Originally posted by @ravi-mosaicml in https://github.com/mosaicml/composer/issues/632#issuecomment-1056988994_

**Comment by hanlint** [comment]
Since we support pytorch schedulers, a user could just construct and pass in a `ChainedScheduler` to our trainer.

**Comment by ravi-mosaicml** [comment]
> Since we support pytorch schedulers, a user could just construct and pass in a `ChainedScheduler` to our trainer.

A PyTorch chained scheduler is not compatible with the composer schedulers.

**Comment by hanlint** [comment]
Ah, I should clarify:

We support pytorch schedulers, so if a user really needed a new combination, they could just pass in `schedulers=ChainedScheduler`. They would not be able to use the `Time` abstraction, but I think that's a minor loss for an edge use case.

New schedulers get introduced very rarely, and writing up a custom scheduler is not too difficult. When we discussed this with @jbloxham , the complexity of any of those above options (and general chainability) is not worth the edge cases. This is consistent with other frameworks (e.g. huggingface and PTL), where they've assembled their own schedulers, and defered more complex cases to pytorch native schedulers.

**Comment by ravi-mosaicml** [comment]
Makes sense, not opposed to leaving this up to the user, but curious to hear @dblalock's thoughts (since he made the original comment on the PR?)

**Comment by dblalock** [comment]
Agree that minimizing complexity should be constraint #1. The main opportunity for possible improvement I saw was just making warmup easy to do without duplicate code each time (and ideally automatically supported, since its utility is mostly orthogonal to the precise scheduler chosen). E.g., 

1) **_Most minimal change:_** having some helper function all of our `whatever_with_warmup_scheduler` functions call. Or, 
2) **_Fairly minimal change:_** having a `warmup: Union[Time, str]`  argument for our composer schedulers.

I observed that factoring out warmup _could_ be special case of a more general chaining mechanism, but I won't argue for one; especially given the points raised above.
